WIP: Monolithic Rewrite

has2k1 opened the pull request:
Covers #283 and touches on stuff discussed in #259 and #221.

This is a complete rewrite of ~~the plotting pipeline~~ nearly everything so as to not to violate the [layered-grammar](http://vita.had.co.nz/papers/layered-grammar.pdf). As such there is a close adherence to the ggplot2 ~~plot building pipeline~~ code base.

It is far from being done, the following need to be addressed:
- [x] scale transformations e.g. `log10, sqrt` (concerns in #283)
- [x] matplotlib subplotting stuff
- [x] Legend
- [x] Rebase and apply changes
- [ ] Theme all elements of the plot
- [ ] Fix qplot
- [ ] Replace test infrastructure with newest from MPL. (At present it is acting up)
- [x] Get it working with Python 3
- [ ] Comprehensive tests
- [ ] Code profiling
- [ ] pep8

**Testing and Contributing**
- Install the rewrite branch, `pip install git+https://github.com/has2k1/ggplot.git@rewrite`
- Please report issues at https://github.com/has2k1/ggplot/issues and pull requests at https://github.com/has2k1/ggplot/pulls
- Thanks

has2k1 committed:
initial commit

has2k1 committed:
Some tests

has2k1 committed:
Add panel and facets

- Introduce plot_build routine that where all the parts of the plot
  are assembled.

has2k1 committed:
Clean up ggplot, add layer, ... run stuff

- Clean up and run plot_build to see how it is working out.

has2k1 committed:
Compute aesthetics

has2k1 committed:
plot_build - Add group

has2k1 committed:
Fix deepcopy and facets

- For geom, stat and layer do not copy the data
- Override the previous faceting implementation

has2k1 committed:
pd.Categorical sorting and train position

- Get the categoricals to sort properly -- making use
  of Pandas upstream bug fixes
- Implement position scale training

has2k1 committed:
Add margins

- Port add_margins from reshape::add_margins
  and tests

has2k1 committed:
Map position, ninteraction

- Map positions scales and deal with scale limits
- Change dataframe_id to ninteraction and reimplement it
  numerically like plyr::id
- Get rid of a bunch of super() calls as they break when
  ggplot is reloaded(key in developing with ipython notebook).
- Clean up the position scales with intermediate
  base classes that handle special cases.
- implement utils.utils.censor
- fix stat.__radd__, how the layer is added. (It is still not
  clean enough)

has2k1 committed:
Add calculate_stats to the plot pipepline

- Plus layer.calc_statistic, stat._calculate_groups
- Fix stat_bin._calculate to make use of the range
- Add range related scale.* $ utility methods
- Split waiver() into a setter and a checker (is_waive())
- some tests

has2k1 committed:
Add map_statistic to the plot pipeline

- Include support for any computed stats
- Reset index for the data that moves through the plotline
- various np.asarray tweaks where necessary
- fix scales.input() to flatten the lists of aesthetics

has2k1 committed:
Add reprameterisation to the plot pipeline

- Implement geom_bar.reparameterise
- Split geom_area and geom_ribbon
- Port resolution() from R

has2k1 committed:
Add adjust_position to the plot pipeline

- position baseclass, position_jitter and stabs for other
  position classes
- gg_import utility function
- change some global ggplot.* imports to relative imports

has2k1 committed:
Implement the position classes that use 'collide'

- positions dodge, stack, fill and jitterdodge

- capture the enviroment of the `ggplot()` call instead
  the `aes()`. There is potential more than one `aes()`
  which leads to a mix up because of aesthetic overrideing

- `layer.compute_aesthetics` now does the transforms

- fix the `range` and `limit()` of the discrete position scales.
  scale ranges are either of type list for discrete scales or
  ndarray for continuous scales. Discrete position scales can
  tell which training was done based on range type, incontrast
  ggplot2 introduces a different name `range_c`

- `pd.pivot_table` depracation of `rows`

- gg_warning utility function to print the warning messages

- remove_missing utility function, it will be used in a number
  of places

has2k1 committed:
Reset scales and train positions again

- Had to adopt the second range for discrete position scales,
  and all ranges are lists.

has2k1 committed:
Train non positions scales and panels ranges

- Train and map non position scales
- Calculate the x & y ranges for each panel
- Fix calls to ninteraction. They used the old call signature

has2k1 committed:
Stats learned to use trained scales

- Get `stats` to use the x & y scale dimensions
- `stats` let calculate_groups do the putting together of the data
- stat_bin2d is now functional
- pd.isnull is friendlier than np.isnan <- freaks out on strings
- cleanup some more global imports
- geoms learned layer_params

has2k1 committed:
Remove old scale code

has2k1 committed:
Draw the plot and facets

- Plotting in the `ggplot`, `layer` and `geom` objects
- Introduce new (possibly still evolving) API i.e
  `draw_group()` and `draw` for the geoms
- Update `geom_point` and use rgba array to avoid
  groupbys on alpha mappings.
- Fix missing `nrow` and `ncol` in the facets
- Draw facet labels

has2k1 committed:
Get themes to handle the panel ticks correctly

- Ticks should be only around the left and bottom edges of the
  left and bottom plots and not all the plots in the figure,

has2k1 committed:
Turn off subplot sharex and sharey

- set sharex and sharey to False, they complicate the managing
  of the axis ticks and ticklabels

has2k1 committed:
Get geoms abline, hline and vline working

- These geoms still have same type of issues as ggplot2.
  https://github.com/hadley/ggplot2/issues/426
  This is a badge-of-honor :)

has2k1 committed:
Get geoms ribbon and area working

- geom_ribbon and geom_area, can map to alpha aesthetic, ggplot2
  does not do this.
- Add huron dataset
- Close any stale plots before drawing another. Works in ipython,
  not tested elsewhere.

has2k1 committed:
midpoint parameter for scale_color_gradient2

- same signature as ggplot2

closes #363

has2k1 committed:
Introduce gg_context and get geom_line working

- Create gg_context a separate context management class, it also
  minimises the clutter in `ggplot.draw`.
- gg_warning is now manages the printed warning messages. The
  context takes cares of when to reset the messages
- Use 'zorder' to make sure that geoms are displayed in the correct
  order.

has2k1 committed:
geom_path with size mapping

- Make geom_line use geom_path
- fix geom.draw_groups to make plot info from grouped data

closes #161

has2k1 committed:
geom_step inherits geom_path & learned 'direction'

closes #294

has2k1 committed:
Fix gg_context and against theme_xkcd transforms

- Save the mpl_context. Was creating and exiting separate contexts.
- Another guard against deep copying theme_xkcd transforms when
  copying rcParams

has2k1 committed:
More relevant apply_theme and some tidying up

- seaborn does not do minor gridlines
- some pep8 touches

has2k1 committed:
Use make_color_tuples around hex_to_rgba

Deal with a mixture of list-like (mappings) and scalar (direct settings)
`facecolor` and `alpha` aesthetics. `make_color_tuples` knows how to
do this.

has2k1 committed:
Update geom_rect

- geom_rect now uses collections to draw the rectangles. This is
  faster and clearer.
- fixed `scale_size.__init__`, the range argument is not a range
  for the scale in that it can be trainned, it is a range for the
  palette.
- Instance of `scale` should be able to initialize and set the
  scale parameters e.g limits.

has2k1 committed:
No more dataframe.groupby().apply()

Pandas calls the function two times on the first grouped dataframe.
This was making it hard to debug, and in cases where the function
calls code that also does a `dataframe.groupby().apply` the
extra calls multiply.

The solution is to do this with a `groupby_apply` utility
function that splits-applys-combines in a straightforward way.

has2k1 committed:
Update geom_bar, plus do breaks and labels

- `geom_bar` uses the `geom_rect.draw` method and for now will inherit,
  `geom_rect`. It may be best to make all the draw methods static, so
  that inheritances (& mixins) aren't used except for the much more
  closely related geoms.

- Handle the breaks and labels, allowing for matplotlib to take care
  of this task automatically if non have been directly set by the user
  or as a result of the manipulations in preparing the plot.

has2k1 committed:
facet_wrap and slightly smaller gridlines

- Get facet_wrap working
- gridlines & axis ticks have the same linewidth

has2k1 committed:
update geom_density

- geom_ribbon knows how to draw lines without fill and fill without
  lines. While `None` as a default paramter value looks closer the
  ggplot2 API, an empty string `''` has the functional equivalent
  i.e No color or No fill

has2k1 committed:
Use statsmodels for the KDE in stat_density

It allows for more options, current status not withstanding.

has2k1 committed:
Silence SettingWithcopyWarnings

has2k1 committed:
Update geom_tile

- It uses geom_rect. It needs some profiling.

has2k1 committed:
Make geom.draw a static method

- This opens up all draw methods which in addition to reparameterising
  allows the draw methods to be building blocks.

has2k1 committed:
Update geom_smooth

has2k1 commented:
**Todo**
- Add test from #409

has2k1 committed:
Upgraded pandas & numpy. API changed, fixes needed

- Category 'levels' are now 'categories'
- Deprecation, pandas indexes cannot use operators (eg. -, +)
- Datetime and timestamp handling is broken (not fixed yet)

has2k1 committed:
Update geom_text, geom_now_its_art

- For geom_text, ggplot infrastructure now takes care of
  the axis dimensions

has2k1 committed:
Match stat_bin to ggplot2

- ported `fullseq()` from Hadleys scales package
- fixed #404

has2k1 committed:
Fix panel order in for facet_grid

Make sure the panel order in the data to be plotted
is consistent with the order in panels dataframe.

has2k1 committed:
Add other computed stats to stat_density

- Also, only the range of the x-scale should be required. Currently,
  it is coming up empty if x is of type datetime. However, that is
  no reason to use the y range.

has2k1 committed:
Environment is captured in aes() call

- Now that every thing is working, we revert the environment
  capture to the aes() call.

has2k1 committed:
Inherit from dict instead of UserDict

- There are less suprises with a straight forward dict.
  e.g set(userdict) has to be set(userdict.keys())

has2k1 committed:
Create guide definitions for the legend

- Create the legend classes. These hold the information required
  to build the legends.
- Almost all other changes are part of the support infrastracture
  for the guide

has2k1 committed:
Introduced legend into the plot

- add some theme features that control some aspects of
  the legend
- implement geom_point.draw_legend

has2k1 committed:
Add geom_path.draw_legend

- Rectangle background for the legend key gets an edge of the same
  color as the foreground since a 1px edge weather transparent or
  not is part the rectangle. Otherwise, lines that go upto and
  touch the edge appear out of bounds.

has2k1 committed:
Add geom_text.draw_legend

- Make sure there is a legend before creating an anchored_box
- scale_size is continuous

has2k1 committed:
Add geom_smooth.draw_legend

- geom.layer_params absorbed into geom.DEFAULT_PARAMS,
  and all is sorted out in geom.__init__
- stat does not directly add layer

has2k1 committed:
Added geom_polygon and it's draw_legend

- It provides a draw_legend for other geoms
- Made sure the group parameter is respected, both as an aesthetic
  and as a layer parameter

has2k1 committed:
Remove 'group' parameter unknown to matplotlib

has2k1 committed:
Add labels to the legend entries

- Also fixed Categorical order so that the labels are correct. This
  was due to an incomplete fix in an earlier commit.

has2k1 committed:
Pass layer to geom.draw_legend

- Through layer, draw_legend has access to the geom and
  stat. This allows for potentially more involved legend
  entries. e.g. geom_polyon

has2k1 committed:
A bunch of small fixes

- legend labels may take from stat default aes, the specifications
  in there are more informative than the geom. This is how gglot
  does it
- make sure legend label entries are strings
- matplotlib API for colors is inconsistent. Using None instead of
  empty string is working better.
- geom_polygon.draw_legend does not vary the linewidth of the
  rectangle.

has2k1 committed:
Legend placeable on all four sides

- Allow for variable number of rows and columns for the guides
  in the legend.

has2k1 committed:
Scale transforms and scale_datetime

- Added scale datetime, smoothers do not mess with the
  date anymore.
- Added transform creating functions i.e xxx_trans
- Fixed some empty facet labels

JanSchulz commented:
@has2k1 Is there anything I can help?

has2k1 commented:
Not yet, I will call out when I get to that point (probably to do with profiling). The rebase will be nasty, I think I will have to rebase and force my version for all commonly altered files.

has2k1 committed:
Sort out breaks, labels & trans for cont. scales

- Continuous scales knows how to transform using the trans
  object. The trans object no wraps up transform and inverse
  functions.
- Legends now display labels in original data space
- Resurrect format functions.
- _waiver_ object does not get deep copied.
- Removed stale code cluttering up `scales/.utils.py`

has2k1 committed:
Do not let mpl default colors sneak in

- When fill or color aesthetics None
- This should be the last blow to this color is
  None, '' or 'none' inconsistency thing.

has2k1 committed:
Tweak facet layout and lenged title

- Don't show empty facets for facet wrap
- Try again for pixel perfect facet labels. This is a fools effort.
- Align legend according to the theme parameter
- Clean up ggsave and use tight bbox by default. It makes sure all
  of the legend (artists) are in the viewport.
- Remove some old code

has2k1 committed:
Fixup facet wrap axes

- The main issue was, locators were applied to axes after the
  breaks and labels were turned off. This had an effect of putting
  the breaks and labels back on.
- Also put all the breaks, labels and limits code in one function
- The axis parameter setting work is no longer offset the theme.

has2k1 committed:
Remove mpl default fc on poly. legend

has2k1 committed:
Clean up

- Use more list comprehensions instead of map
  and filter.
- Remove obsolute code.
- Make geom.reparameterize a staticmethod.
- Proper inheritance especially for color scales,
  make sure the super class `__init__` method is
  is called.
- Deal with the axis ticks and label explicitly.
- Refactor continuous color scales to use `gradient_n_pal`.
- Use the `warnings` module
- Use floating point ordinals for datetime in the scale utils.

has2k1 committed:
Fix the adding of default scales

- When a stat computes a column and that columns needs
  a new scale, the logic of figuring out the column was
  off.
- Enable retransforms after stat computation
- Fix a slip up in the previous clean up commit

has2k1 committed:
Rebased and seem successful

This is the first commit after a big rebase using the command

`git rebase --strategy=recursive --strategy-option=theirs upstream/master`

The effect of this is that `upstream/master` is clobbered in case of
merge conflicts and only new files have survived. The branches had
diverged at commit 29ec5abad4b620f575c12f96eef914595524f28d. With the
rebase overlaying a near rewrite not much has been lost in terms
functionality, however that related to the new files that have come in
is going to be integerated. i.e the new color scheme and geoms. Although
not verified, it is very likely that some new test cases have not survived.

If you are tracking the changes and are reading this, I am remorseful
for any undue suffering that my actions have caused.

In this commit obsolete import statements have been removed, this gets
everything functional.

JanSchulz commented on ggplot/geoms/geom.py:
```diff
@@ -199,7 +209,7 @@ def _find_aes_and_data(self, args, kwargs):
         for arg in args:
             if isinstance(arg, aes) and passed_aes:
-                raise Execption(aes_err)
+                raise Exception(aes_err)
```
GgplotException?

has2k1 replied:
There are still a couple of those lying around and some have been fixed in my current branch. I plan to grep for all errors and make sure they are the right type.

JanSchulz commented on ggplot/geoms/geom_now_its_art.py:
```diff
@@ -1,8 +1,10 @@
 from __future__ import (absolute_import, division, print_function,
```
IMO this whole file can go... :-)

JanSchulz commented on ggplot/geoms/geom_step.py:
```diff
@@ -1,31 +1,31 @@
-    REQUIRED_AES = {'x', 'y'}
-    DEFAULT_PARAMS = {'stat': 'identity', 'position': 'identity',
-            'direction': 'hv'}
+# TODO: Add test case
```
yes, please :-)

JanSchulz commented on ggplot/ggplot.py:
```diff
@@ -122,417 +84,389 @@ class _empty(object):
 
         return result
 
-    def draw(self):
```
I used draw in some SO questions. It would be nice if that could at least be kept with a warning and call the function which now does the job (render?)

JanSchulz commented on ggplot/ggplot.py:
```diff
@@ -122,417 +84,389 @@ class _empty(object):
+        """
+        plt.close("all")
+        with gg_context(theme=self.theme):
+            plot = self.draw()
```
Oups. so draw is still there -> I think this should be renamed so that draw is still the highlevel function which calls the lower ones.

has2k1 replied:
I have not yet come to a conclusion with those two functions. There is a certain consistency I want with similarly named methods across multiple modules. That will definitely put draw at the top and make it part of the user API.

JanSchulz commented on ggplot/ggplot.py:
```diff
@@ -122,417 +84,389 @@ class _empty(object):
+        """
+        Render the complete plot and return the matplotlib figure
+        """
+        plt.close("all")
```
IMO `plt.close("all")` should not be used at all by library code: if do some matplotlib first and then use ggplot, my manual figures are all gone...

My opinion is also that all 'plt' usage should be gone and figures/axis (.> The OO interface of matplotlib) should be used directly. If one is to use gridspec, this makes it possible to pass in a gridspec slice instead of a figure and get "pasting together" of two ggplots in one figure for free :-)

See here for the idea: https://github.com/mwaskom/seaborn/issues/399#issuecomment-84035872

has2k1 replied:
It already bit me, and it will be gone. I just didn't like plot spillage in ipython notebook when developing and there are errors.

JanSchulz replied:
the inline backend (`%matplotlib inline`) already does a `plt.close("all")` at the end of a cell if I remember correctly...

has2k1 replied:
It was not doing it for me when I added that line to code. But that could have been with ipython `1.*`

JanSchulz commented on ggplot/ggplot.py:
```diff
@@ -122,417 +84,389 @@ class _empty(object):
+        apply_facet_spacing(plot)
+        return plot
+
+    def plot_build(self):
```
build_plot?

has2k1 replied:
Makes more sense, but for some of those functions I named them similar to ggplot2 since the code closely follows it. In case there is a bug, I want to make it easy to cross correlate the two code bases.

JanSchulz commented on ggplot/scales/__init__.py:
```diff
@@ -1,30 +1,122 @@
 
-__all__ = ['scale_color_brewer', 'scale_colour_brewer',
+# color
+from .scale_color import scale_color_brewer, scale_colour_brewer
```
Somehow this long list screams for a way to not duplicate the list. Probably like adding doing a `{<from>:['items',...],...}` and then do some [import magic](https://docs.python.org/2/library/functions.html#__import__) to load the thing and put it in `__all__` in one small nested loop...

JanSchulz commented on ggplot/scales/scale_color.py:
```diff
@@ -0,0 +1,278 @@
+        # Only draw the maximum allowable colors from the palette
+        # and fill any remaining spots with None
+        _n = n if n <= nmax else nmax
+        bmap = brewer2mpl.get_map(palette_name, type, _n)
```
brewer2mpl has a namechange, so should get a conditional import :-(

https://github.com/jiffyclub/palettable/wiki

JanSchulz commented on ggplot/tests/__init__.py:
```diff
@@ -24,32 +26,34 @@ def teardown_package():
-    'ggplot.tests.test_theme_gray',
-    'ggplot.tests.test_theme_mpl',
-    'ggplot.tests.test_theme_seaborn'
+    # 'ggplot.tests.test_basic',
```
:-(

has2k1 replied:
As I add new tests, I don't want the full test suite running.

JanSchulz commented on ggplot/utils/ggutils.py:
```diff
@@ -2,25 +2,87 @@
+from .exceptions import gg_warning, GgplotError, gg_reset
+
+
+if not hasattr(mpl, 'rc_context'):
```
This will error as utils has that removes. Also probably not needed anymore with gg_context?

JanSchulz commented on ggplot/utils/ggutils.py:
```diff
@@ -2,25 +2,87 @@
+
+
+if not hasattr(mpl, 'rc_context'):
+    from .utils import _rc_context
```
This will error on mpl 1.1.1 as utils has that removes. Also probably not needed anymore with gg_context?

JanSchulz commented:
@has2k1 Can you do a rebase so that travis can do it's work again? This PR is quite big and I fear that merging it will be quite hard, so makeing it any easier would be good

-> rebase (and maybe squash together into more logical units), so that one can see the difference and also get travis running again?
-> enable the unittests so that one can be sure that this doesn't break (too much)
-> build more unittests?

It would probably also be good to get this into a branch in the default repo, so that others can try it more easily -> @glamp, what do you say?

has2k1 commented:
I did try a rebase in the last 24 hours and it seems to have gone well enough. However, there is still some more work to be done before this PR is ready for merging. The goal is for the final thing to have no loose ends. 

When I decided it would more or less be a complete rewrite, I pretty much worked on whatever I wanted (or what irked me at the time). Unfortunately I am still doing the same, so I would not count on "squashable" logical units. That should change as it stabilizes and activity gets driven more by bugs and features than by "plenty of interconnected stuff to work on".

The existing unittests will be enabled in due time. As the plot layout improves and everything gets fine tuned I don't want to be swapping "expected" images all the time, and without committing to that effort the tests are not really helpful. On the general testing front I am more concerned about the internals and how to test/verify that the plotting information has been assembled properly.

It will be ready for users to try out sooner when I am confident of not getting barraged with too many complaints. Need to figure out a way to get testers.


For completeness, we still need to straighten out the documentation, requirements and a proper release system.

has2k1 committed:
Use color system from seaborn and rm old files

- This is the color system that was used by upstream at rebase time.
- Remove obsolete files.
- Make GgplotError handle tuple of strings.
- .gitignore is aware of a few more .devstuff directoreis

has2k1 committed:
Fix up gg_warning, ggplot.draw & Exceptions

- Removed obsolete warning/Exception advisory message
- waiver object is know print friendly

has2k1 committed:
Some refactoring and restructuring

- Move all ported methods from R scales package to the
  `scales/utils.py` file
- Fix weird warnings.warn error. It does not like `category=None`
  as an argument even though that is the default acc. to the python
  source.
- Move code to and from `components/`. Either way, what belongs
  in there is not immediately obvious. Thought, maybe create a
  third-party directory.

has2k1 committed:
Add missing datasets that are in ggplot2.

has2k1 committed:
Shrink plot information if it is similar

Instead of passing a list(s) of similar values in the plot
information, shrink it into a scalar. In some cases, this helps
matplotlib make better choices for drawing calls.

has2k1 committed:
geom_path, linejoin, lineend and arrow

- geom_path has learned about linejoin, lineend and arrow
  parameters
- arrow usage matches ggplot2 as close as possible
- fix error in geom_line where only 'x' and 'y' aesthetics
  where being sorted

has2k1 committed:
Update stat_summary, fix line stats & update tests

- stat_summary summarizes on the group and x so all the
  calculations happen in _calculate_groups.
- Drop the fudge factors for positioning the facet labels.
  The QT Backend does not need any fudging. png and pdf need
  different values for pixel perfect positions.
- Made sure stat function work
- Other touchups on stats
- Removed obsolete tests directory

has2k1 committed:
Add labels and title components

- Modify xlab, ylab, labs, ggtitle, xlim and ylim to fit the
  new infrastructure and put them under in the components
  sub-package.
- Capitalize some globals.

has2k1 committed:
Cleanup trans and scale_datetime

- Remove duplicate code from scale_datetime and made sure
  scale_continuous takes care of it.
- Some transformation functions where missing return values.
- Formatters in trans are now factories for consistency with
  the Locators

has2k1 committed:
Add coord_cartesian

- The `scales` and the `coordinates` are now separate. `scale.coord_*`
  methods are no more.
- The coord_cartesian implementation tries to mirror that of ggplot2.
  A consequence of this shift is that the `scales` parameter pass to
  `geom.draw_groups` and `geom.draw` methods just has the range
  information. They are not `scale` instances as one would first
  expect. At this point the instances were used to transform the data
  and then compute the range information.

has2k1 committed:
Fix issue with origin adjustments in stat_bin

- Allows for proper centering of the bars
- Implement a rough equivalent to Rs `seq()` to help prevent off-by
  "delta" errors when generating sequences that should include the
  right bound.
- Fixes #427
- Closes #429

JanSchulz commented:
So what's the status here? If this isn't getting merged soon, we we have a problem, as ggplot is unuseable with latest pandas (https://github.com/yhat/ggplot/issues/417#issuecomment-114313988, http://stackoverflow.com/questions/31911321/geom-histogram-in-python-ggplot-does-not-work-on-ipython-3) 

See also this: https://github.com/winpython/winpython/issues/109#issuecomment-123401141

cc: @glamp

has2k1 commented:
Given that I have not fixed the tests as yet, it is not ready for master. In addition, some `geoms` that are usable  in master are still not up-to-date in my branch.

Also, because we have a number of upstream dependencies we need a clearer policy/rationale on which versions to support, especially Pandas and Matplotlib. Pandas has had consequential API changes for 0.15.* and 0.16.0 releases and will probably continue to do so in the short to medium term.  Plus, starting with Matplotlib 2.0 there will be bigger changes(API or otherwise) and we will have to keep up.

More important, going forward we need to adhere to the *grammar of graphics* philosophy -- so PRs need to solve issues the right way otherwise the project ends up in a dead zone.

JanSchulz commented:
So, when will this be ready? I really like the stuff I see in this PR and I also think the "grammar of graphics" stuff is awesome, but the way this change is landing in ggplot is hurting, as our users basically see this as a dead project after no proper release in a year :-(

Re API changes in upstream projects: mpl will probably be fine, they have a very strict API policy. Going with the styles API would be nice, as this lets us probably do some changes more easily... Pandas is actually kind of our fault: they had two major versions which deprecated this, so almost a year of warnings... :-(

has2k1 commented:
I can't tell when it will be ready, the upside is it will be close to parity with ggplot2.

I have reluctantly given this out

```
pip install git+https://github.com/has2k1/ggplot.git@rewrite
```

to a handful of people to get some limited feedback, but when the tests and Python 3 support are fixed I will solicit wider feedback from `r/python` and maybe the MPL user mailing list.

eco32i commented:
@has2k1 So I checked out the rewrite branch.

Issues I have had so far:
* [error caused by wrong method name in EvalEnvironment in `components/layer.py`](https://github.com/has2k1/ggplot/pull/1)

naught101 commented:
It would be nice if a new pull request from has2k1/rewrite could be made - then progress could be tracked from here.

glamp commented:
@has2k1 Would it be possible to just fix what you have and merge in? I realize it's not done, but I'm worried it will never be completed if we hold out for the feature complete, "same as ggplot2" version.

has2k1 commented:
@glamp, I cannot really do that, so far issues that have seemed postponable have later on turned out to be main cause of nasty surprises. The audience would lose trust in the package.

Either way it is a lot closer since the last run of commits. I think I will do one more and then solicit for some wider-scale testing.

jan-glx commented:
Hi @has2k1 ,
great to see that you are bringing ggplot for python back to life. I was just using your branch but experienced issues when plotting DataFrames with 5 rows due to the way `pinfo` is impelemented/defined. I honestly think its entries should be either of `len()` `1` or `n` e.g. pinfo['linestyle'] should be `["solid"]` not `"solid"` as the latter has `len()` `5` as well. But I am afraid that this needs changes in a lot pf places... Btw. did you consider making this a branch in yhat/ or master in has2k1/ and allowing issues? The current situation feels kind of discouraging to possible participants. All the best,
Jan

````
import pandas as pd
from ggplot import *

df= pd.DataFrame({'x': list(range(5)),'y': list(range(5))})
ggplot(aes(x='x', y='y'), data=df) +  stat_smooth(color='red')
# gives: ValueError: Do not know how to convert [u's', u'o', u'l', u'i', u'd'] to dashes
````

has2k1 commented:
@jan-glx, I have fixed the issue. I have made some changes to the first post on how to contribute.

On the `["solid"]` issue, it is an [optimization](https://github.com/has2k1/ggplot/blob/rewrite/ggplot/geoms/geom.py#L316-L326).

jan-glx commented:
@has2k1: awesome.

machow commented:
Hey all, I've checked up on this ggplot branch a few times before noticing this rewrite. It looks like @has2k1 would like to wait to finish making changes, and more power to him, since the contribution looks really impressive.

However, if I wanted to contribute to the project, I would be faced with two options..

1. clone the yhat/ggplot master, and hope that what I'm changing hasn't already been addressed here.
2. clone this branch, and work with @has2k1. In this case, it's basically like having an alternative (albeit temporary?) ggplot, maintained by @has2k1.

In this case, what would the best course of action be? I'm asking, because I feel like this branch is hovering ambiguously between a permanent fork of the fairly inactive yhat/ggpot project, and an active but informal "dev" branch of the yhat/ggplot. My feeling is that it would be easier to contribute if the activity of this branch being pulled was kept on a yhat/ggplot dev branch, so people could easily find it and check it out. This would also address @has2k1's concerns with putting incomplete or buggy code on master.

The work in this pull request looks great--seriously, kudos to @has2k1 for pushing things along!

JanSchulz commented:
The first post was updated by @has2k1 to show the way :-).

I would suggest to start with the testing framework and the tests (see here: https://github.com/has2k1/ggplot/blob/rewrite/ggplot/tests/__init__.py#L28 most of them are currently disabled). This is mostly enabling one test, running `python tests.py` and checking if the generated images fit to the baseline images and if not (which will mostly be the case due to changes to e.g. the legend code) check that the new images make sense and then update the baseline images. And if the generated images make no sense report this as errors.

JanSchulz commented:
@has2k1 I find the idea to push your branch to a dev branch in this repo quite good. This would alos eliminate the problem of two bugtrackers (only use this one) and will probably get travis support out of the box.

naught101 commented:
Any chance of a status update @has2k1?

has2k1 commented:
@naught101, my *rewrite* branch can be installed, I try not to break anything in there -- it is not my development branch. I just completed a long overdue (due to Pandas limitations) re-factoring of the `geom`s and with that the way data flows from input to plotting has been tidied up. This should allow me a second go at flippable (`coord_flip`) x-y coordinates and also make it easier for others to write `geom`s.

I also have a branch with new documentation (minus any plots at the moment) and I am getting closer to revamping the tests. Getting these two out of the way should bring about open season.

machow commented:
If this is how yhat/ggplot wants to do things, I think @glamp and @janschulz need to seriously consider putting an immediately visible link to @has2k1's branch (at the top of the README or repo description). This refactor is going to be one giant pull request, without discussion on individual components via issues? I'll take what I can get, and appreciate the time @has2k1 has put into fixing up the library, but the main branch does a poor job of conveying what is going on, and how it prefers to manage issues.

@janschulz mentioned that I should file issues on @has2k1's branch, but someone would have to stumble upon this pull request to get that information.

janschulz commented:
@machow I'm not anymore involved

naught101 commented:
Wow, what just happened? @glamp, you just re-wrote the whole package from scratch in a month or two?

eco32i commented:
So what happened to @has2k1 rewrite branch? Is it a fork now?

has2k1 commented:
I guess it is.

hoechenberger commented:
I bit more and open communication would be very much appreciated :)
